src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -754,106 +759,252 @@ public static boolean bothSingleByteOptimizable(DynamicObject string, DynamicObj
     }
 
     @CoreMethod(names = "count", rest = true)
-    @ImportStatic(StringGuards.class)
     public abstract static class CountNode extends CoreMethodArrayArgumentsNode {
 
-        @Child private EncodingNodes.CheckEncodingNode checkEncodingNode = EncodingNodesFactory.CheckEncodingNodeGen.create(null, null);
         @Child private ToStrNode toStr = ToStrNodeGen.create(null);
+        @Child private CountRopesNode countRopesNode = CountRopesNode.create();
+
+        @Specialization(guards = "args.length == size", limit = "getCacheLimit()")
+        public int count(VirtualFrame frame, DynamicObject string, Object[] args,
+                @Cached("args.length") int size) {
+            Rope[] ropes = argRopes(frame, args);
+            return countRopesNode.executeCount(string, ropes);
+        }
+
+        @Specialization(replaces = "count")
+        public int count(VirtualFrame frame, DynamicObject string, Object[] args) {
+            Rope[] ropes = argRopesSlow(frame, args);
+            return countRopesNode.executeCount(string, ropes);
+        }
+
+        @ExplodeLoop
+        protected Rope[] argRopes(VirtualFrame frame, Object[] args) {
+            Rope[] strs = new Rope[args.length];
+            for (int i = 0; i < args.length; i++) {
+                strs[i] = rope(toStr.executeToStr(frame, args[i]));
+            }
+            return strs;
+        }
+
+        protected Rope[] argRopesSlow(VirtualFrame frame, Object[] args) {
+            Rope[] strs = new Rope[args.length];
+            for (int i = 0; i < args.length; i++) {
+                strs[i] = rope(toStr.executeToStr(frame, args[i]));
+            }
+            return strs;
+        }
+
+        protected int getCacheLimit() {
+            return getContext().getOptions().DEFAULT_CACHE;
+        }
+    }
+
+    @ImportStatic(StringGuards.class)
+    public abstract static class CountRopesNode extends TrTableNode {
+
+        public static CountRopesNode create() {
+            return CountRopesNodeFactory.create(null);
+        }
+
+        public abstract int executeCount(DynamicObject string, Rope[] ropes);
 
         @Specialization(guards = "isEmpty(string)")
         public int count(DynamicObject string, Object[] args) {
             return 0;
         }
 
+        @Specialization(guards = { "!isEmpty(string)", "cachedArgs.length > 0",
```

Comment:
I don't know why, but it looks a bit odd to me to have a guard on `cachedArgs.length > 0` instead of `args.length > 0`. Is there a reason to prefer the cached one?